Add skipValues config option

milkaknap opened the pull request:
I've used serializeJSON in such use case, when I don't want empty strings to be serialized and sent to server. I want to skip them. Just serialize string fields which values are not empty strings. 

So I've implemented more generic solution. There is new **skipValues** option available. It's key-value dict, where _key_ is any of available (also custom) value types and _value_ is list of "skip values" (values, which should be skipped during serialization).

I've also updated tests and doc. 

Please let me know what you think about this pull request.

marioizquierdo commented:
This feature is nice. Most of the time, the server can deal with empty strings and such, but it's true that some times filtering those things on the client is easier and convenient.

This problem can be solved in different ways, like using the `parseWithFunction` option, or with jQuery by disabling the fields before serializing:

```javascript
$('form').find(':input[value=""]').attr('disabled', true);
$('form').serializeJSON();
$('form').find(':input[value=""]').attr('disabled', false);
```
Although it's not as declarative and easy as having an option just for that.

Another way to solve this would be to have a data attribute to filter empty values for the most common cases, like `data-omit-falsy=true` that would skip falsy values like "", null, 0. This can also be applied per field which is more useful than applying per type.

However, since you already implemented the skip-per-type feature, that made sense for you, and since it doesn't add much complexity to the Readme file, I think is ok to include this option.

But have some change requests:
  * Change the option name to `skipValuesForType`
  * Please add a test case, at least, for the example you used in the README file.
  * In the README, you use the example of "number: [0, null]", but I don't think a `:number` field can be parsed as null value. It should be "number: [0, NaN]" instead (because `Number("null")` returns NaN). But this has another problem; in JavaScript `NaN === NaN` is `false`, so in reality not valid numbers can not be skip with this technique. I don't want to add an option that can have surprising behavior. See if you can come up with something that works here, otherwise it's better to change the option for `skipFalsyValues: bool` or `skipFalsyValuesForType:{}`, implemented as just checking if the parsed value is truthy or fasly.

milkaknap commented:
@marioizquierdo: Thank's for your review! 

I have deleted the option `skipValues` and added new `skipFalsyValuesForTypes` and `skipFalsyValuesForFields` options also with `data-skip-falsy` input attribute. I think this can be more usefull.

I've created some test cases too.

Please let me know, if anything else should be updated.

marioizquierdo commented on spec/spec.js:
```diff
@@ -1013,6 +1093,7 @@ describe("$.serializeJSON", function () {
 
     describe('with modified defaults', function() {
       var defaults = $.serializeJSON.defaultOptions;
+      console.log(defaults);
```
Remove this log statement

marioizquierdo commented on jquery.serializejson.js:
```diff
@@ -194,6 +214,21 @@
+      if (skipFalsyFromDataAttr && (skipFalsyFromDataAttr === "true" || skipFalsyFromDataAttr == "false"))
+        return (skipFalsyFromDataAttr === "true")
+      return null;
+    },
```
Please add braces to the `if (...) { }` statement to make my linter happy :)

marioizquierdo replied:
I would change the logic to: `return skipFalsyFromDataAttr !== "false")`, so anything works as a true value except for the explicit "false" that is meant to override the general option. This is better even if the documentation says to only use "true" or "false", because sometimes people makes typos or they prefer to use "on" instead of "true", in which case it's better to assume a true value than ignoring it (IMO). This will also make it work if the attribute is specified without a value (i.e. `<input data-skip-falsy>` should be assumed enabled).

```js
      if (skipFalsyFromDataAttr != null) {
        return skipFalsyFromDataAttr !== "false";
      }
      return null;
```

marioizquierdo commented on jquery.serializejson.js:
```diff
@@ -45,7 +45,22 @@
+            skipSerialization = true;
+          else if (opts.skipFalsyValuesForFields && opts.skipFalsyValuesForFields.indexOf(nameWithNoType) >= 0)
+            skipSerialization = true;
+        }
```
Please move this logic to a helper function, and use braces to make my linter happy :)

marioizquierdo commented on README.md:
```diff
@@ -220,6 +220,8 @@ To change the default behavior you use the following options:
   * **parseAll: true**, all of the above. This is the same as if the default :type was `:auto` instead of `:string`.
   * **parseWithFunction: function**, define your own parse function(inputValue, inputName) { return parsedValue }
+  * **skipFalsyValuesForTypes**, to skip serialization of falsy values for listed types
+  * **skipFalsyValuesForFields**, to skip serialization of falsy values for listed field names. It can be set (or overriden) by setting `data-skip-falsy` input attribute to `true` or `false`.
```
Add default values `[]`, and a little example if possible::

```
**skipFalsyValuesForTypes: []**, skip fields with given types and falsy values (i.e. `skipFalsyValuesForTypes: ["string", "number"]` would skip `""` for `:string` fields, and `0` for `:number` fields).
**skipFalsyValuesForFields: []**, skip fields with given names and falsy values. You can use `data-skip-falsy="true"` input attribute as well. Falsy values are determined after converting to a given type, note that `"0"` as :string is truthy, but `0` as :number is falsy.
```

marioizquierdo commented on README.md:
```diff
@@ -477,6 +479,13 @@ obj = $form.find(':input').filter(function () {
+ * `skipFalsyValuesForFields` option to skip empty values for listed field names
+ * `data-skip-falsy="true"` input field attribute to skip empty values on the current field
+
+_Note: To make it work correctly you should use it on fields with :type or set the `parseAll` option to `true`._
```
No need to explain the options here again, you can just make a quick to the explanation before:

```
For falsy values (`false, "", 0, null, undefined, NaN`), you can use serializeJSON specific options to skip fields by name (`skipFalsyValuesForFields: [name]`) or by type (`skipFalsyValuesForTypes: ["string", "null"]`).

You can also use a data attribute `data-skip-falsy="true"` on the inputs that should be ignored (`data-skip-falsy` is aware of field :types, so it knows how to skip a non-empty input like this `<input name="foo:number" value="0" data-skip-falsy="true">`).
```

milkaknap commented:
@marioizquierdo: Thanks for the code review! Hope to have corrected all the mistakes.

I think, skip by type can be used well with the custom types. For example, I've created `emptyString` type to be used with the all fields which should be skipped when empty.

marioizquierdo commented:
I made a refactor/cleanup pass after merging this here: https://github.com/marioizquierdo/jquery.serializeJSON/commit/1cff5e45bfc77c5bb6291344b717bcb4c1837291

I added a few extra test cases to cover possible combinations like when doing `skipFalsyValuesForTypes: ["string"]`, if a field has no type, it should be assumed to be a string, or checking that it works when mixing `data-value-type` and `data-skip-falsy` together.

I refactored the code a little bit to reuse the function that finds an input by type and attr.

And I added a breaking change: Empty arrays are not considered "falsy" (you had a test case to check for this). I hope this doesn't cause you any problem.

I will now bump the library version and release the new options.

milkaknap commented:
OK, thanks!
